lib/jerboa/client/worker.ex

```diff
@@ -4,70 +4,166 @@ defmodule Jerboa.Client.Worker do
+  end
+  def handle_call(:allocate, _, state) do
+    if state.relayed_address do
+      {:reply, {:ok, state.relayed_address}, state}
+    else
+      {result, new_state} = request_allocation(state)
+      case result do
+        {:ok, _} ->
+          {:reply, result, update_lifetime_timer(new_state)}
+        _ ->
+          {:reply, result, new_state}
+      end
+    end
+  end
+  def handle_call(:refresh, _, state) do
+    unless state.relayed_address do
+      {:reply, {:error, :no_allocation}, state}
+    else
+      {result, new_state} = request_refresh(state)
+      case result do
+        :ok ->
+          {:reply, result, update_lifetime_timer(new_state)}
+        _ ->
+          {:reply, result, new_state}
+      end
+    end
   end
-  def handle_call(:persist, _, state) do
-    msg = Jerboa.Format.encode(Jerboa.Params.put_class(binding_(), :indication))
-    {:reply, cast(socket(state), server(state), msg), state}
+
+  def handle_cast(:persist, state) do
+    state
+    |> Protocol.bind_ind()
+    |> send_ind()
+    {:noreply, state}
+  end
+
+  def handle_info(:allocation_expired, state) do
+    new_state = %{state | lifetime_timer_ref: nil,
+                          relayed_address: nil,
+                          lifetime: nil}
+    {:noreply, new_state}
   end
 
   def terminate(_, state) do
     :ok = UDP.close(socket(state))
   end
 
-  defp server(%__MODULE__{server: s}) do
-    s
-  end
+  ## Internals
+
+  defp server(%{server: server}), do: server
 
-  defp socket(%__MODULE__{socket: s}) do
-    s
+  defp socket(%{socket: socket}), do: socket
+
+  @spec send_req(state) :: state
+  defp send_req(%{transaction: %{req: req} = transaction} = state) do
+    socket = socket(state)
+    {address, port} = server(state)
+    :ok = UDP.send(socket, address, port, req)
+    {:ok, {^address, ^port, response}} = UDP.recv(socket, 0, timeout())
+    %{state | transaction: %{transaction | resp: response}}
   end
 
-  defp binding_ do
-    Params.new()
-    |> Params.put_method(:binding)
+  @spec send_ind(state) :: :ok
+  defp send_ind(%{transaction: %{req: msg}} = state) do
+    socket = socket(state)
+    {address, port} = server(state)
+    :ok = UDP.send(socket, address, port, msg)
   end
 
-  defp reflexive_candidate(params) do
-    xor_mapped_addr = Params.get_attr(params, XORMappedAddress)
-    {xor_mapped_addr.address, xor_mapped_addr.port}
+  defp timeout do
+    Keyword.fetch!(Application.fetch_env!(:jerboa, :client), :timeout)
   end
 
-  defp call(socket, {address, port}, request) do
-    :ok = UDP.send(socket, address, port, request)
-    {:ok, {^address, ^port, response}} = UDP.recv(socket, 0, timeout())
-    response
+  @spec request_allocation(state) :: {result :: term, state}
+  defp request_allocation(state) do
+    {result, new_state} =
+      state
+      |> Protocol.allocate_req()
+      |> send_req()
+      |> Protocol.eval_allocate_resp()
+    case result do
+      :retry ->
+        request_allocation(new_state)
```

Comment:
I think that we need to distinguish situations when we want to retry because there's no response (cases when we suspect network failure) from the cases where according to the protocol we could retry.

In the first case the counter looks ok to me (providing the retires are at least logged).